framework/TSA/RWD.py

```diff
@@ -0,0 +1,244 @@
+      SignatureMatrix = np.zeros((SignatureWindowLength, AllWindowNumber))
+      for i in range(AllWindowNumber):
+        SignatureMatrix[:,i] = np.copy(history[i:i+SignatureWindowLength])
+
+      # Sequential sampling
+      if SampleType == 0:
+        BaseMatrix = np.copy(SignatureMatrix)
+
+      # Randomized sampling
+      elif SampleType == 1:
+        sampleLimit = len(history)-SignatureWindowLength
+
+        WindowNumber = sampleLimit//4
+        sampleIndex = np.random.randint(sampleLimit, size=WindowNumber)
+        BaseMatrix = np.zeros((SignatureWindowLength, WindowNumber))
+        for i in range(WindowNumber):
+          WindowIndex = sampleIndex[i]
+          BaseMatrix[:,i] = np.copy(history[WindowIndex:WindowIndex+SignatureWindowLength])
+      # Piecewise Sampling
+      else:
+        WindowNumber = len(history)//SignatureWindowLength
+        BaseMatrix = np.zeros((SignatureWindowLength, WindowNumber))
+        for i in range(WindowNumber-1):
+          BaseMatrix[:,i] = np.copy(history[i*SignatureWindowLength:(i+1)*SignatureWindowLength])
+      U,s,V = mathUtils.computeTruncatedSingularValueDecomposition(BaseMatrix,0)
+      FeatureMatrix = U.T @ SignatureMatrix
+      params[target] = {'UVec'   : U[:,0:fi],
+                        'Feature': FeatureMatrix}
+    return params
+
+
+  def getParamNames(self, settings):
+    """
+      Return list of expected variable names based on the parameters
+      @ In, settings, dict, training parameters for this algorithm
+      @ Out, names, list, string list of names
+    """
+    names = []
+    for target in settings['target']:
+      base = f'{self.name}__{target}'
+      sw = int(settings['SignatureWindowLength'])
+      fi = int(settings['FeatureIndex'])
+      for i in range(fi):
+        for j in range(sw):
+          names.append(f'{base}__UVec{i}_{j}')
+
+    return names
+
+  def getParamsAsVars(self, params):
+    """
+      Map characterization parameters into flattened variable format
+      @ In, params, dict, trained parameters (as from characterize)
+      @ Out, rlz, dict, realization-style response
+    """
+    rlz = {}
+    for target, info in params.items():
+      base = f'{self.name}__{target}'
+      (k,l) = (info['UVec']).shape
+      for i in range(l):
+        for j in range(k):
+          rlz[f'{base}__UVec{i}_{j}'] = info['UVec'][j,i]
+    return rlz
+
+
+
+  def generate(self, params, pivot, settings):
+    """
+      Generates a synthetic history from fitted parameters.
+      @ In, params, dict, characterization such as otained from self.characterize()
+      @ In, pivot, np.array(float), pivot parameter values
+      @ In, settings, dict, additional settings specific to algorithm
+      @ Out, synthetic, np.array(float), synthetic estimated model signal
+    """
+
+    synthetic = np.zeros((len(pivot), len(params)))
+    for t, (target, _) in enumerate(params.items()):
+      SigMat_synthetic = params[target]['UVec'] @ params[target]['Feature']
```

Comment:
change 'SigMat_synthetic' to 'sigMatSynthetic', underscore is not allowed, please check RAVEN python format https://github.com/idaholab/raven/wiki/RAVEN-Software-Coding-Standard
Please check and fix all the violations